### PR TITLE
Folding overflowing intrinsics with select

### PR DESCRIPTION
**Pull request by lialan**

### CODE REVIEW

**Comment by lialan** [comment]
okay ... I have to change base branch to `dev-15` to trigger CI. But it is not supposed to be merged directly. It should be merged to #225

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.004 ║
║ Best                                6.452 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.002 ║
║ Best                               11.111 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.001 ║
║ Best                                7.477 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.024 ║
║ Best                                0.964 ║
║ Worst                               0.000 ║
║ Total                               0.034 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.038 ║
║ Best                                1.152 ║
║ Worst                               0.000 ║
║ Total                               0.134 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.002 ║
║ Best                                0.141 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.004 ║
║ Best                                6.452 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.002 ║
║ Best                               11.111 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.001 ║
║ Best                                7.477 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.024 ║
║ Best                                0.964 ║
║ Worst                               0.000 ║
║ Total                               0.034 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.038 ║
║ Best                                1.152 ║
║ Worst                               0.000 ║
║ Total                               0.134 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.002 ║
║ Best                                0.141 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.004 ║
║ Best                                6.452 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.002 ║
║ Best                               11.111 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.001 ║
║ Best                                7.477 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.024 ║
║ Best                                0.964 ║
║ Worst                               0.000 ║
║ Total                               0.034 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.038 ║
║ Best                                1.152 ║
║ Worst                               0.000 ║
║ Total                               0.134 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.002 ║
║ Best                                0.141 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.004 ║
║ Best                                6.452 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.002 ║
║ Best                               11.111 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.001 ║
║ Best                                7.477 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.024 ║
║ Best                                0.964 ║
║ Worst                               0.000 ║
║ Total                               0.034 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.038 ║
║ Best                                1.152 ║
║ Worst                               0.000 ║
║ Total                               0.134 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.002 ║
║ Best                                0.141 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Target/SyncVM/SyncVMExpandSelect.cpp [hunk 1]
```diff
@@ -81,6 +81,8 @@ bool SyncVMExpandSelect::runOnMachineFunction(MachineFunction &MF) {
       {SyncVMCC::COND_LE, SyncVMCC::COND_GT},
       {SyncVMCC::COND_GT, SyncVMCC::COND_LE},
       {SyncVMCC::COND_GE, SyncVMCC::COND_LT},
+      // COND_OF is an alias for COND_LT
```
Comment (akiramenai) [inline llvm/lib/Target/SyncVM/SyncVMExpandSelect.cpp:84]: It's not quite the case. Before start folding we only had flags from sub, and LT is the flag set by sub. We introduced OF to tell different semantic as LLVM doesn't expect flags to have different meaning depending on which instruction set them. For instance, I worry about mul, as the spec says:
```
◦ LT_OF is set if overflow occurs, i.e. op1×op2≥2
◦ EQ is set if resultlow=0.
◦ GT is set if LT_OF and EQ are cleared.
```
So GT doesn't seem to be a proper reversal of OF, 0 is the corner case.
Can we have an execution test to cover that nuances?

File: llvm/test/CodeGen/SyncVM/overflow.ll [hunk 2]
```diff
@@ -1202,3 +1203,40 @@ exit:
...
+  ret i256 %selected
+}
+
+; CHECK-LABEL: mul_test
+define i256 @mul_test(i256 %a, i256 %b, i256 %x, i256 %y) {
+entry:
+  %res1 = call {i256, i1} @llvm.umul.with.overflow.i256(i256 %x, i256 %y)
+; CHECK:      mul!    r3, r4, r{{[0-9]+}}, r{{[0-9]+}}
+; add.ge is add.of reversed
+; CHECK-NEXT: add.ge  r2, r0, r1
```
Comment (akiramenai) [inline llvm/test/CodeGen/SyncVM/overflow.ll:1262]: Can we have a test checking for lt/of, not the reversal?

**Review by akiramenai — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/SyncVM/SyncVMExpandSelect.cpp (hunk 1 above)
Comment (lialan) [reply llvm/lib/Target/SyncVM/SyncVMExpandSelect.cpp:84]: @akiramenai We introduced OF because we do not want passes earlier in the pipeline to reverse it into GE. Here at this very late stage, I think flipping is okay provided that no other alterations after it.

BTW, I think you meant GE instead of GT? LT_OF's reverse is GE.

File: llvm/test/CodeGen/SyncVM/overflow.ll (hunk 2 above)
Comment (lialan) [reply llvm/test/CodeGen/SyncVM/overflow.ll:1262]: done. added test to show `.of` instructions generated.

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.004 ║
║ Best                                6.452 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.002 ║
║ Best                               11.111 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.001 ║
║ Best                                7.477 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.024 ║
║ Best                                0.964 ║
║ Worst                               0.000 ║
║ Total                               0.034 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.038 ║
║ Best                                1.152 ║
║ Worst                               0.000 ║
║ Total                               0.134 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.002 ║
║ Best                                0.141 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.004 ║
║ Best                                6.452 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.002 ║
║ Best                               11.111 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.001 ║
║ Best                                7.477 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.024 ║
║ Best                                0.964 ║
║ Worst                               0.000 ║
║ Total                               0.034 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.038 ║
║ Best                                1.152 ║
║ Worst                               0.000 ║
║ Total                               0.134 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.002 ║
║ Best                                0.141 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Target/SyncVM/SyncVMExpandSelect.cpp (hunk 1 above)
Comment (akiramenai) [reply llvm/lib/Target/SyncVM/SyncVMExpandSelect.cpp:84]: @lialan I see... with the spec in coq it looks like we can reverse OF, can't we? I mean our OF which can only be set by an overflowing instruction. All instruction that can set OF allow to use GE as its reverse.
If so, why do we want earlier passes to avoid reversing in that case? And (not in scope of this patch) it looks that we can reverse OF in combine flag setting pass. Am I right?

**Review by akiramenai — APPROVED** [review]
Please squash.

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.004 ║
║ Best                                6.452 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.002 ║
║ Best                               11.111 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.001 ║
║ Best                                7.477 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.024 ║
║ Best                                0.964 ║
║ Worst                               0.000 ║
║ Total                               0.034 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.038 ║
║ Best                                1.152 ║
║ Worst                               0.000 ║
║ Total                               0.134 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.002 ║
║ Best                                0.141 ║
║ Worst                               0.000 ║
║ Total                               0.003 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Target/SyncVM/SyncVMExpandSelect.cpp (hunk 1 above)
Comment (lialan) [reply llvm/lib/Target/SyncVM/SyncVMExpandSelect.cpp:84]: @akiramenai There was a reason specifically for that, pertaining to machine block placement pass reversing OF to GE and cause havoc. We had a discussion on this, and Shamatar was there too. Can't remember the details, perhaps we discussed it on the now-retired mattermost.

let me get back to you if I find anything.
